File: kirun-js/src/engine/json/schema/convertor/exception/SchemaConversionException.ts
import { ConversionMode } from '../enums/ConversionMode';

export class SchemaConversionException extends Error {
    private readonly schemaPath: string;
    private readonly source?: string | null;
    private readonly mode?: ConversionMode | null;
    cause?: Error;

    constructor(
        schemaPath: string,
        source: string,
        message: string,
        mode?: ConversionMode,
        sce: SchemaConversionException[] = [],
        err?: Error,
    ) {
        super(message + (sce ? sce.map((e) => e.message).reduce((a, c) => a + '\n' + c, '') : ''));
        this.schemaPath = schemaPath;
        this.source = source ?? null;
        this.mode = mode ?? null;
        this.cause = err;
    }

    public getSchemaPath(): string {
        return this.schemaPath;
    }

    public getSource(): string | null {
        return this.source ?? null;
    }

    public getMode(): ConversionMode | null {
        return this.mode ?? null;
    }

    public getCause(): Error | undefined {
        return this.cause;
    }
}

File: kirun-py/src/kirun_py/json/schema/convertor/exception/schema_conversion_exception.py
from __future__ import annotations
from typing import Optional, Any, TYPE_CHECKING

if TYPE_CHECKING:
    from kirun_py.json.schema.convertor.enums.conversion_mode import ConversionMode


class SchemaConversionException(Exception):
    def __init__(
        self,
        schema_path: str,
        source: Any = None,
        mode: Optional[ConversionMode] = None,
        cause: Optional[Exception] = None,
        message: str = '',
    ):
        super().__init__(
            message or f'Schema conversion failed at {schema_path}'
        )
        self._schema_path = schema_path
        self._source = source
        self._mode = mode
        self._cause = cause

    def get_schema_path(self) -> str:
        return self._schema_path

    def get_source(self) -> Any:
        return self._source

    def get_mode(self) -> Optional[ConversionMode]:
        return self._mode

    def get_cause(self) -> Optional[Exception]:
        return self._cause
